digital_land/utils/add_endpoints_utils.py

```diff
@@ -0,0 +1,708 @@
+    :return:
+    """
+    endpoint_path = ctx.obj["COLLECTION_DIR"] / "endpoint.csv"
+    collection_dir_str = ctx.obj["COLLECTION_DIR"].absolute()
+    pipeline = ctx.obj["PIPELINE"]
+
+    collect(
+        endpoint_path,
+        collection_dir_str,
+        pipeline,
+    )
+
+
+def task_populate_resource_and_log_csvs(ctx):
+    collection = ctx.obj["COLLECTION"]
+    collection_dir = ctx.obj["COLLECTION_DIR"]
+
+    logs_path = collection_dir / "log/*/"
+    collection = load_log_items(collection, log_directory=logs_path)
+
+    collection.log.save_csv(directory=collection.directory)
+    collection.resource.save_csv(directory=collection.directory)
+
+
+def task_generate_lookup_entries(ctx):
+    """
+    uses the dataset entity ID ranges declared in specification, and collected
+    resources to generate new entries for the pipeline's lookup.csv
+    :param ctx:
+    :return:
+    """
+
+    pipeline_dir = ctx.obj["PIPELINE_DIR"]
+    pipeline = ctx.obj["PIPELINE"]
+    dataset_name = pipeline.name
+    collection = ctx.obj["COLLECTION"]
+    collection_resource_dir = ctx.obj["COLLECTION_RESOURCE_DIR"]
+    organisation_dir = ctx.obj["ORGANISATION_DIR"]
+    specification = ctx.obj["SPECIFICATION"]
+    specification_dir = ctx.obj["SPECIFICATION_DIR"]
+    tmp_dir = ctx.obj["TMP_DIR"]
+
+    # get organisation.csv
+    copy_successful = copy_latest_organisation_files_to(organisation_dir)
+
+    if copy_successful:
+        # create and initialise entity number generator
+        entity_num_state = get_entity_number_state(
+            dataset=dataset_name,
+            pipeline_dir=pipeline_dir,
+            specification_dir=specification_dir,
+        )
+
+        entity_num_gen = EntityNumGen(entity_num_state=entity_num_state)
+
+        print("")
+        print("======================================================================")
+        print("New Lookups")
+        print("======================================================================")
+
+        # generate the lookup entries for each new resource
+        dataset_resource_map = collection.dataset_resource_map()
+        all_lookup_entries = []
+        for dataset in dataset_resource_map:
+            for resource in dataset_resource_map[dataset]:
+                resource_file_path = collection_resource_dir / resource
+
+                resource_lookups = get_resource_unidentified_lookups(
+                    input_path=resource_file_path,
+                    dataset=dataset,
+                    organisations=collection.resource_organisations(resource),
+                    pipeline=pipeline,
+                    specification=specification,
+                    tmp_dir=tmp_dir.absolute(),
+                    org_csv_path=(organisation_dir / "organisation.csv").absolute(),
+                    entity_num_gen=entity_num_gen,
+                )
+
+                all_lookup_entries += resource_lookups
+
+        ctx.obj["NEW_LOOKUPS"] = all_lookup_entries
+
+
```

Comment:
I'm not sure passing this volume of objects through ctx is a good idea. Makes the code hard to read and your never too sure what should be in the ctx